fee_calculator/apps/api/serializers.py

```diff
@@ -8,6 +8,59 @@
 )
 
 
+class SchemeListQuerySerializer(serializers.Serializer):
+    type = serializers.ChoiceField(help_text="Graduated fee scheme type",
+                                   choices=('AGFS', 'LGFS'),
+                                   required=False)
+    case_date = serializers.CharField(help_text="Date for which you would like a list of applicable fee schemes",
+                                      required=False)
+
+
+class BasePriceFilteredQuerySerializer(serializers.Serializer):
+    scenario = serializers.IntegerField(help_text='',
+                                        required=False,)
+    advocate_type = serializers.CharField(help_text='''Note the query will return prices with `advocate_type_id`
+                                                        either matching the value or null.''',
+                                          required=False,)
+    offence_class = serializers.CharField(help_text='''Note the query will return prices with `offence_class_id`
+                                                        either matching the value or null.''',
+                                          required=False,)
+    fee_type_code = serializers.CharField(help_text='',
+                                          required=False,)
+
+
+class CalculatorQuerySerializer(serializers.Serializer):
+    scenario = serializers.IntegerField(help_text='',
+                                        required=True,)
+    fee_type_code = serializers.CharField(help_text='',
+                                          required=True,)
+    advocate_type = serializers.CharField(help_text='''Note the query will return prices with `advocate_type_id`
+                                                        either matching the value or null.''',
+                                          required=False,)
+    offence_class = serializers.CharField(help_text='''Note the query will return prices with `offence_class_id`
+                                                        either matching the value or null.''',
+                                          required=False,)
+
+    def __init__(self, *args, **kwargs):
+        for unit in Unit.objects.all():
+            f"""{unit.pk.lower()} = serializers.DecimalField(
```

Comment:
Huh, I'm surprised that even works! Without the `exec()` I'm surprised it does anything at all...

I'll be honest, it isn't quite what I meant, as it's the `exec("this string of code that really wanted to not be run dynamically")` pattern I didn't like.

I'm not deeply familiar with Django Rest Framework (DRF), but from my limited experience and by analogy with other similar Django concepts (largely form classes) I would expect to be something more like `self.fields[unit.pk.lower() = serializers.DecimalField()`. Not sure on the specifics of this, but just to illustrate that what I'm thinking is rather than trying to copy the way we'd write it if we were defining the class, we instead accept that that ship has sailed (indeed, even if we were to `exec()` the Field definition onto the instance, it may be that what DRF does with it will already have happened) and go to the end result which is the fields dictionary, and make our changes there. As I said originally, though, this may be difficult for some reason, and it might not be soon enough for us if we need them on the class for some reason.

[The DRF docs](https://www.django-rest-framework.org/api-guide/serializers/#dynamically-modifying-fields) (anchor link isn't going to the right part of the page for me) do talk about modifying the `fields` dict, but only to pop fields out rather than add new ones, so I'm still not sure tbh

Reply:
Thanks, that makes a lot of sense. Unfortunately it looks like the `fields` dict on a serializer is immutable. Any attempt to amend it after the instance has been created (beyond `pop`ping unwanted fields out, as in the example you shared) seems to fail. Which probably explains why Joel did it the way he did.

The easy fix seems to be to hard-code all the `unit` and `modifier_type` fields into the serializer in the same way the `scenario`, `fee_type_code` etc fields are, instead of iterating over the `Unit` and `ModifierType` objects and dynamically adding fields. This makes the class pretty big and clunky and adds some duplication but avoids `exec`ing anything. 

eg. 
```
class CalculatorQuerySerializer(serializers.Serializer):
    scenario = serializers.IntegerField(help_text='',
                                        required=True,)
    fee_type_code = serializers.CharField(help_text='',
                                          required=True,)
    advocate_type = serializers.CharField(help_text='Note the query will return prices with `advocate_type_id` either matching the value or null.',
                                          required=False,)
    offence_class = serializers.CharField(help_text='Note the query will return prices with `offence_class_id` either matching the value or null.',
                                          required=False,)
    case = serializers.DecimalField(help_text='Quantity of the price unit: Case',
                                    required=False,
                                    max_digits=100,
                                    decimal_places=5)
    day = serializers.DecimalField(help_text='Quantity of the price unit: Whole Days',
                                   required=False,
                                   max_digits=100,
                                   decimal_places=5)
    defendant = serializers.DecimalField(help_text='Quantity of the price unit: Defendants',
                                         required=False,
                                         max_digits=100,
                                         decimal_places=5)

...
```

I'm sure there's a cleverer way of doing it but I can't find it 😬. I'll do some more thinking (ie searching stackoverflow) and see if I can come up with a cleaner solution.

Reply:
Ah yes, hadn't thought about it being immutable. Makes sense, and at least is better than it being mutable and then us running into unexpected behaviour later :)

Only thing I can think to suggest at this point is that if you wanted to reduce the repetition a bit you could subclass the `DecimalField` to already have some of those values you're always passing it (required=False, max_digits, decimal_places), so you'd still have a line per field and hard-coding but it would look more like `defendant = ModifierField(help_text='Quantity of the price unit: Defendants')` or even `defendant = ModifierField('Defendants')`


But if that's not good enough, we may be able to find something marginally better (more robust, anyway) than the original, perhaps by going into the class's `__new__`, no real idea specifically but that feels like it could be a thing